### ISSUE TITLE
Location not being saved/read

### ISSUE THREAD

**Issue by pmdevita** [pmdevita/StayPut#6]
From @edcous

```
>.... [16:40:08 INFO]: [StayPut] PlayerTeleportEvent activated
>.... [16:40:08 INFO]: [StayPut] Player teleporting: myusername
>.... [16:40:08 INFO]: [StayPut] Player teleporting from world: world1
>.... [16:40:08 INFO]: [StayPut] Player teleporting to world: world2
>.... [16:40:08 INFO]: [StayPut] Saving myusername (myusername)'s location in >world1<
>.... [16:40:08 INFO]: [StayPut] Not teleporting player because there is no previous location found 
```

and from RoboTitanII on spigotmc.org

```
[21:19:35 INFO]: [StayPut] PlayerTeleportEvent activated
[21:19:35 INFO]: [StayPut] Player teleporting: Username
[21:19:35 INFO]: [StayPut] Player teleporting from world: world1
[21:19:35 INFO]: [StayPut] Player teleporting to world: world2
[21:19:35 INFO]: [StayPut] Saving Username (UUID)'s location in >world1<
[21:19:36 INFO]: [StayPut] Not teleporting player because there is no previous location found
```

Inferring that this happens when teleporting back from lobby to world1, this seems to point to an issue either with the saving or loading of the location from the DB. I'm guessing this is an SQLite issue since I haven't tested that as much but it's still unconfirmed.

@edcous, when you switched to MySQL did it start working?

**Comment by edcous** [comment]
No

**Comment by edcous** [comment]
I’m thinking of switching back to multiverse inventories, because it seems that works in 1.15.

**Comment by pmdevita** [comment]
If that works best for you then go for it. But if you give me some time, I'm sure I can fix the issue.

**Comment by edcous** [comment]
I don't know what I did, but somehow I installed multiverse inventories and perworldinventories at the same time and it works.

**Comment by pmdevita** [comment]
RoboTitan was apparently using MySQL so it looks like the issue isn't backend related. 

Just tested it on 1.15.2 with SQLite, works for me. If anyone else is experiencing the issue, let me know.

**Comment by edcous** [comment]
Idk what I did, but somehow it works now.

**Comment by pmdevita** [comment]
1006c61026be94125e3a8e4a4602c5b5bf030af6/1.3.2 may have fixed this issue or a subset of those experiencing it. Keeping issue open for a little until I don't get any more reports with the issue.

**Comment by edcous** [comment]
Cool, I’ll probably backup my server and see if that fixes it.

**Comment by edcous** [comment]
It works!!!! Thank you so so much!!!

**Comment by pmdevita** [comment]
You're welcome!

Awaiting confirmation from one more user on spigotmc.org and then closing.

**Comment by edcous** [comment]
are you going to close this?

**Comment by pmdevita** [comment]
Haven't heard back from him, issue seems to have been resolved.

**Comment by edcous** [comment]
Just wondering, what ended up being the problem?

**Comment by pmdevita** [comment]
This bug appeared with the support of generic teleport events.

All teleports are now handled through the generic teleport event (they go through there anyways at some point, makes things a little simpler). But the problem with listening to this is that there are a lot of different reasons somebody could be teleporting and the plugin needs to decide which teleports to redirect.

There are two kinds of teleports the plugin has interest in, which I have dubbed "directed" and "non-directed" teleports. A non-directed teleport is simply a request to be moved to a specific world (ex. `/mv tp playername worldname`). A directed teleport is a request to move worlds and then go to a specific location in that world (ex. `/tp playerone playertwo` [with playertwo being in a different world]). When someone performs a non-directed teleport, they should be sent to their last location. When someone performs a directed teleport, the should be sent to the location in the world they are trying to go. 

Unfortunately, whether or not a teleport is directed is not explicitly given in the teleport event. Instead, it must be inferred from the location data. A non-directed teleport uses the location of the spawn of the world and thus, if a teleport is using the world spawn it's a non-directed teleport. If it is anywhere other than world spawn, then it is directed.

Finally, we get to the problem. When the plugin was checking against the spawn of the world, it was using the normal Spigot API methods for acquiring that information. However, if one used Multiverse to change the spawn of a world, non-directed teleports would use the Multiverse set spawn, which is different from Spigot's. Since these two locations didn't match, the plugin assumed it was a directed teleport and not redirect it, sending you to the Multiverse set spawn. 

The solution was to use the world spawn from the Multiverse API instead of the Spigot API, as it would always give the right answer (even if spawn was not set with Multiverse). It's a simple fix but discovering there was a difference and pointing the blame at Multiverse took a bit of testing as I was unaware the Spigot API would ever be incorrect about something like that.

tl;dr StayPut thought people weren't being sent to spawn and didn't redirect them, turns out Multiverse mucks up where spawn is

Additional side note: You may wonder what happens if someone attempts a directed teleport to spawn. Unfortunately, due to the lack of info in the teleport event, there is no way to differentiate between a non-directed teleport and a directed teleport aimed at a world spawn. The plugin will treat both as non-directed and send the player to their last location. There is some extra work that can be done by reading the multiverse events that fire before the actual teleport event but it's a little complicated and won't completely solve the problem. The solution is to make the destination ever so slightly different from the world spawn (even changing head orientation works) or do a second jump to spawn after entering the world (the plugin doesn't intervene for same world teleports).

**Comment by edcous** [comment]
Wow, this all seems super complicated. I’m starting to learn java, so hopefully this will make more sense soon.

**Comment by pmdevita** [comment]
Don't worry, everyone has to start somewhere. Also, most of the work for this fix was really in the theory, the actual change was maybe a line or two of code.

**Comment by edcous** [comment]
Thanks, maybe I can help improve it when I know more :)
